yosh/shell.py

```diff
@@ -9,59 +13,70 @@
 
 
 def tokenize(string):
-    return shlex.split(string)
+    token=shlex.split(string)
+    for i, el in enumerate(token):
+        if el.startswith('$'):
+           token[i] = os.getenv(token[i][1:])           
+    return token
 
+def handler_kill(signum, frame):
+    raise OSError("Killed!")
 
 def execute(cmd_tokens):
+  if cmd_tokens:
     # Extract command name and arguments from tokens
     cmd_name = cmd_tokens[0]
     cmd_args = cmd_tokens[1:]
 
     # If the command is a built-in command, invoke its function with arguments
     if cmd_name in built_in_cmds:
         return built_in_cmds[cmd_name](cmd_args)
-
-    # Fork a child shell process
-    # If the current process is a child process, its `pid` is set to `0`
-    # else the current process is a parent process and the value of `pid`
-    # is the process id of its child process.
-    pid = os.fork()
-
-    if pid == 0:
-        # Child process
-        # Replace the child shell process with the program called with exec
-        os.execvp(cmd_tokens[0], cmd_tokens)
-    elif pid > 0:
-        # Parent process
-        while True:
-            # Wait response status from its child process (identified with pid)
-            wpid, status = os.waitpid(pid, 0)
-
-            # Finish waiting if its child process exits normally or is
-            # terminated by a signal
-            if os.WIFEXITED(status) or os.WIFSIGNALED(status):
-                break
-
-    # Return status indicating to wait for next command in shell_loop
-    return SHELL_STATUS_RUN
+    global sh
+    signal.signal(signal.SIGINT,handler_kill)
+    # Written in beautiful sentences to run the command,modifed by Ted
```

Comment:
- Do not need to specify who is a modifier in code.

- Change comment to be more specific what it does. (Like `Spawn a child process`)